### DISCUSSION TITLE
Setting Tab order fails

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [tui-cs/Terminal.Gui#2867] [Q&A] [answered]
I am using version 1.14.0 in a PowerShell script. I am struggling to get the tab order to work. From the documentation, I can create this control.

```powershell
$txtDescription = [Terminal.Gui.TextField]@{
    X        = $lblDescription.Frame.Width + 2
    Width    = 40
    Y        = 4
    TabIndex = 2
    TabStop  = $True
    CanFocus = $True
}
```

If my first TabIndex is 0, I expect this control to be the third tab stop. But it isn't. In fact, if I create this control in the console as a standalone object, the tab index is 0.

```powershell
PS C:\> $txtDescription | select tab*,can*

TabIndexes TabIndex TabStop CanFocus
---------- -------- ------- --------
{}                0    True     True
```

I assume I can assign a value. But it doesn't work, and there are no errors.

```powershell
PS C:\> $txtDescription.TabIndex = 2
PS C:\> $txtDescription.TabIndex
0
```

Am I missing something regarding setting tab stops?

**Comment by BDisp** [comment]
The `TabIndex` handles the `TabIndexes` and it can't be greater than the `TabIndexes.Count - 1`. So if you have only one view that can be focused the  `TabIndex` will be equal to zero.

**Reply by jdhitsolutions** [reply]
Bear with me, as I don't have a developer background, so some of the concepts here elude me. I have one top window which contains several controls like text fields and buttons. I'd like to define a tab order among some of the controls. Are you saying this is not possible with a single window? As it stands now, I can set focus to my first control and tab between controls, but I can't seem to control the order.

**Reply by BDisp** [reply]
No problem. You can set the `TabIndex` you want because the `TabIndexes` list is independent from the `Subviews` list. They have the same size but could have different order. What I referred is if you set a `TabIndex` greater than `TabIndexes.Count -1` it will set with the maximum of `TabIndexes.Count -1`. The `TabIndexes` list will be updated to reflect the change and ensuring unique values. Also there the `TabStop` where you enable or disable if a control can get the focus when navigating between the controls.

**Reply by jdhitsolutions** [reply]
I think my head is still spinning from this. Here are the first 3 controls in my preferred tab order.

```powershell
$txtTaskName = [Terminal.Gui.TextField]@{
    X        = $lblDescription.Frame.Width + 2
    Width    = 31
    Y        = 2
    TabIndex = 0
    TabStop  = $True
    CanFocus = $True
}
$dropCategory = [Terminal.Gui.ListView]@{
    Width                   = 15
    Height                  = 8
    X                       = $lblCategory.Frame.Right + 2
    Y                       = $lblCategory.Y
    TabIndex                = 1
    TabStop                 = $True
    AllowsMultipleSelection = $False
    CanFocus                = $True
}
$txtDescription = [Terminal.Gui.TextField]@{
    X        = $lblDescription.Frame.Width + 2
    Width    = 40
    Y        = 4
    TabIndex = 2
    TabStop  = $True
    CanFocus = $True
}
# add each control to top window
```

The first control is set to have focus when the window opens. Tabbing does not move in the expected order. As far as I know I'm not intentionally doing anything with `Subviews`. I may have invalid assumptions about how this is supposed to work.

**Reply by BDisp** [reply]
The controls are only added to the `TabIndexes` list when they are adding to the `SuperView`. I forgot to say that we can't set the `TabIndex` on control instance but only after they are added to the `SuperView`, sorry about that. Then if the `TabIndex` of one control isn't the desired you can change the order by setting;
```pwsh
# after add each control to top window
...
$txtDescription.TabIndex = 0
$dropCategory.TabIndex = 1
$txtDescription.TabIndex = 2
```

**Reply by jdhitsolutions** [reply]
I actually think I follow this. I'll give it a try.

**Reply by jdhitsolutions** [reply, marked as answer]
Finally worked this out.  I realized I had to add the window to the top, which I'm assuming is what creates the `SuperView`.

This code sets the tab order I was after.
```powershell
    $controls.foreach({ $window.Add($_) })

    [Terminal.Gui.Application]::Top.Add($window)
    #set tab order
    $txtTaskName.TabIndex = 0
    $dropCategory.TabIndex = 1
    $txtDescription.TabIndex = 2
    $txtDays.TabIndex = 3
    $radioGrp.TabIndex = 4
    $chkWhatIf.TabIndex = 5
    $btnAdd.TabIndex = 6
    $btnSet.TabIndex = 7
    $btnComplete.TabIndex = 8
    $btnRemove.TabIndex = 9
```

Thank you for your patience.
